source/symbols/AllTypes.cpp

```diff
@@ -311,7 +311,62 @@ const Type& EnumType::fromSyntax(Compilation& compilation, const EnumTypeSyntax&
         previous = ev.getValue();
         previousRange = ev.getInitializer()->sourceRange;
 
-        checkValue(previous, previousRange.start());
+        if (!previous)
+            return;
+        auto loc = previousRange.start();
+        auto& value = previous.integer();
+
+        // An enumerated name with x or z assignments assigned to an enum with no explicit data type
+        // or an explicit 2-state declaration shall be a syntax error.
+        if (!base->isFourState() && value.hasUnknown()) {
+            scope.addDiag(diag::EnumValueUnknownBits, loc) << value << *base;
+            return;
+        }
+
+        // Any enumeration encoding value that is outside the representable range of the enum base
+        // type shall be an error. For an unsigned base type, this occurs if the cast truncates the
+        // value and any of the discarded bits are nonzero. For a signed base type, this occurs if
+        // the cast truncates the value and any of the discarded bits are not equal to the sign bit
+        // of the result.
+        if (value.getBitWidth() > bitWidth) {
+            logic_t msb = value[int32_t(bitWidth) - 1];
+            auto diff = value.getBitWidth() - bitWidth;
+            bool good = true;
+            if (!value.hasUnknown()) {
+                if (base->isSigned() && exactlyEqual(msb, logic_t(true)))
+                    good = value.countLeadingOnes() >= diff;
+                else
+                    good = value.countLeadingZeros() >= diff;
+            }
+            else {
+                if (!base->isSigned())
+                    msb = logic_t(false);
+                for (auto i = value.getBitWidth() - 1; i >= bitWidth; i--) {
```

Comment:
This seems like it could be pretty slow, accessing individual bits isn't particularly fast. Maybe this whole block of code should be a method on SVInt, and it can look at the words instead of individual bits?

Reply:
I implement two new methods in SVInt.
```c++
    /// Returns true if all bits in [bitWidth-1:msb] are exactly equal.
    bool isSignExtendedFrom(bitwidth_t msb) const;

    /// If bit msb is nonzero, duplicate it to all bits in [bitWidth-1:msb] .
    void signExtendFrom(bitwidth_t msb);
```
The first one simplifies the EnumValueOutOfRange check
```c++
            if (!base->isSigned()) {
                good = exactlyEqual(value[int32_t(bitWidth)], logic_t(false)) &&
                       value.isSignExtendedFrom(bitWidth);
            }
            else
                good = value.isSignExtendedFrom(bitWidth - 1);
```
The second one replace the code in SVInt::
```c++
   // Pretend our width is just the width we shifted to, then signExtend
    return lshr(amount).trunc(contractedWidth).sext(bitWidth);
```
which allocates/deallocates two extra new SVInt's, with more efficient, sign extension in-place
```c++
    auto tmp = lshr(amount);
    tmp.signExtendFrom(contractedWidth - 1);
    return tmp;
```